src/mesh/analysis.rs

```diff
@@ -193,14 +193,57 @@ pub fn border_vertex_indices(edge_sharing: &EdgeSharingMap) -> HashSet<u32> {
     border_vertices
 }
 
+#[derive(Debug, PartialEq)]
+pub enum BorderEdgeLoopsResult {
+    Found(Vec<Vec<OrientedEdge>>),
+    FoundWithNondeterminism(Vec<Vec<OrientedEdge>>),
+    Watertight,
+}
+
 /// Finds continuous loops of border edges, starting from a random edge.
 ///
 /// The mesh may contain holes or islands, therefore it may have an unknown
-/// number of edge loops. If two or more edge loops meet at a single vertex, it
-/// is undefined whether they end up in the result as separate border edge
-/// loops, or are joined into a single border edge loop.
-pub fn border_edge_loops(edge_sharing: &EdgeSharingMap) -> Vec<Vec<OrientedEdge>> {
+/// number of edge loops. If two or more edge loops meet at a single vertex, the
+/// result is non-deterministic: a random path of more possibilities is chosen
+/// to form an edge loop.
+pub fn border_edge_loops(edge_sharing: &EdgeSharingMap) -> BorderEdgeLoopsResult {
     let mut border_edges: Vec<_> = border_edges(edge_sharing).collect();
+
+    // If there are no border edges, the mesh is watertight.
+    if border_edges.is_empty() {
+        return BorderEdgeLoopsResult::Watertight;
+    }
+
+    // Check if the border edges are branching (the loops are touching at a
+    // single vertex). If they are, the result is non-deterministic.
+
+    // FIXME: @Optimization: use a Map structure or at least trim the list of
+    // vertices also from below. Find the maximum index that occurs in the
+    // border edge loop.
+    let max_vertex_index = border_edges
+        .iter()
+        .fold(0, |max, edge| max.max(edge.vertices.0.max(edge.vertices.1)));
+    // A helper vector - each index represents a vertex in the mesh. The items
+    // represent the number of occurrences of each respective vertex in the
+    // border edge list.
+    let mut vertex_occurrence_counts: Vec<u32> = vec![0; cast_usize(max_vertex_index) + 1];
+    // Count the vertex orccurences
+    for edge in &border_edges {
+        vertex_occurrence_counts[cast_usize(edge.vertices.0)] += 1;
+        vertex_occurrence_counts[cast_usize(edge.vertices.1)] += 1;
+    }
+
+    // For a deterministic result, each vertex should occur in the list of
+    // boundary edges exactly twice - in two edges, which meet at the vertex.
+    // Higher number means that more edge loops meet at one vertex and therefore
+    // the edge loop list is non-deterministic and possible wrong. If the vertex
+    // appears only once, the mesh may be erroneous. There are holes in the list
+    // of vertex occurrence, therefore some items may be zero.
+    // FIXME: Panic if the occurrence is 1
```

Comment:
I'd change this FIXME to `FIXME: Examine situation when occurrence is 1 and determine what to do`. For now, we can just assume non-determinism in this case too, although IIRC that situation didn't seem possible from our explorations.

Reply:
i can't think of a scenario where a vertex would be contained only in a single edge. the same applies to a border edge. i think we can assume if this number is 1, there must be some error in our code.

Reply:
If the data comes from a mesh. You can currently hand-craft edge sharing maps like this. This is what I meant by the comment, that we should only be able to construct edge sharing from a mesh. Once we have that, We can start asserting things.